src/main/java/com/commercetools/sync/sdk2/services/impl/BaseTransformServiceImpl.java

```diff
@@ -87,47 +183,52 @@ protected CompletableFuture<Void> fetchAndFillReferenceIdToKeyCache(
                         .build())
             .map(variables -> GraphQLRequestBuilder.of().query(query).variables(variables).build())
             .collect(Collectors.toList());
+    return graphQLRequests;
+  }
 
-    return collectionOfFuturesToFutureOfCollection(
-            graphQLRequests.stream()
-                .map(graphQLRequest -> getCtpClient().graphql().post(graphQLRequest).execute())
-                .collect(Collectors.toList()),
-            Collectors.toList())
-        .thenApply(
-            graphQlResults -> {
-              graphQlResults.stream()
-                  .map(r -> r.getBody().getData())
-                  // todo: set limit to -1, the payload will have errors object but what to do with
-                  // it ?
-                  //                  .filter(Objects::nonNull)
-                  .forEach(
-                      data -> {
-                        ObjectMapper objectMapper = JsonUtils.getConfiguredObjectMapper();
-                        final JsonNode jsonNode = objectMapper.convertValue(data, JsonNode.class);
-                        final Iterator<JsonNode> elements =
-                            jsonNode.get(requestType.getName()).get("results").elements();
-                        while (elements.hasNext()) {
-                          JsonNode idAndKey = elements.next();
-                          fillReferenceIdToKeyCache(
-                              idAndKey.get("id").asText(), idAndKey.get("key").asText());
-                        }
-                      });
-              return null;
-            });
+  protected void cacheResourceReferenceKeys(
+      @Nonnull final List<ApiHttpResponse<GraphQLResponse>> graphQLResults) {
+    GraphQlQueryResource[] queryResources = GraphQlQueryResource.values();
+    for (int i = 0; i < queryResources.length; i++) {
+      cacheResourceReferenceKeys(graphQLResults, queryResources[i]);
+    }
   }
 
-  @Nonnull
-  protected Set<String> getNonCachedReferenceIds(@Nonnull final Set<String> referenceIds) {
-    return referenceIds.stream()
-        .filter(
-            id ->
-                (!referenceIdToKeyCache.containsKey(id)
-                    || KEY_IS_NOT_SET_PLACE_HOLDER.equals(referenceIdToKeyCache.get(id))))
-        .collect(toSet());
+  protected void cacheResourceReferenceKeys(
+      @Nonnull final List<ApiHttpResponse<GraphQLResponse>> graphQLResults,
+      @Nonnull final GraphQlQueryResource requestType) {
+    graphQLResults.stream()
+        .map(r -> r.getBody())
+        .filter(Objects::nonNull)
+        .map(body -> body.getData())
+        // todo: set limit to -1, the payload will have errors object but what to do with
```

Comment:
What do you mean by this comment? `set limit to -1` - which limit is it?

Reply:
Actually, this comment is not mine. I try to clarify with the author.

Reply:
Author isn't available anymore. I've deleted the comment: [0a10f89](https://github.com/commercetools/commercetools-sync-java/pull/941/commits/0a10f899e51f18f3f806e269c5f314299531ce6c)